pkg/instrumentation/podmutator.go

```diff
@@ -159,22 +292,50 @@ func (pm *instPodMutator) Mutate(ctx context.Context, ns corev1.Namespace, pod c
 		logger.Error(err, "failed to select an OpenTelemetry Instrumentation instance for this pod")
 		return pod, err
 	}
-	insts.Sdk = inst
+	insts.Sdk.Instrumentation = inst
 
-	if insts.Java == nil && insts.NodeJS == nil && insts.Python == nil && insts.DotNet == nil && insts.Go == nil && insts.ApacheHttpd == nil && insts.Sdk == nil {
+	if insts.Java.Instrumentation == nil && insts.NodeJS.Instrumentation == nil && insts.Python.Instrumentation == nil &&
+		insts.DotNet.Instrumentation == nil && insts.Go.Instrumentation == nil && insts.ApacheHttpd.Instrumentation == nil &&
+		insts.Sdk.Instrumentation == nil {
 		logger.V(1).Info("annotation not present in deployment, skipping instrumentation injection")
 		return pod, nil
 	}
 
 	// We retrieve the annotation for podname
-	var targetContainers = annotationValue(ns.ObjectMeta, pod.ObjectMeta, annotationInjectContainerName)
+	if featuregate.EnableMultiInstrumentationSupport.IsEnabled() {
+		// We use annotations specific for instrumentation language
+		insts.Java.Containers = annotationValue(ns.ObjectMeta, pod.ObjectMeta, annotationInjectJavaContainersName)
+		insts.NodeJS.Containers = annotationValue(ns.ObjectMeta, pod.ObjectMeta, annotationInjectNodeJSContainersName)
+		insts.Python.Containers = annotationValue(ns.ObjectMeta, pod.ObjectMeta, annotationInjectPythonContainersName)
+		insts.DotNet.Containers = annotationValue(ns.ObjectMeta, pod.ObjectMeta, annotationInjectDotnetContainersName)
+		insts.Go.Containers = annotationValue(ns.ObjectMeta, pod.ObjectMeta, annotationInjectGoContainersName)
+		insts.ApacheHttpd.Containers = annotationValue(ns.ObjectMeta, pod.ObjectMeta, annotationInjectApacheHttpdContainersName)
+		insts.Sdk.Containers = annotationValue(ns.ObjectMeta, pod.ObjectMeta, annotationInjectSdkContainersName)
+
+		// We check if provided annotations and instrumentations are valid
+		ok, msg := insts.areContainerNamesConfiguredForMultipleInstrumentations()
```

Comment:
I may have misunderstood the situation here.

Users should be allowed to enable the feature gate but still use `instrumentation.opentelemetry.io/inject-java: "true"` and provide no container names.

Users should not be allowed to enable the feature gate and use `instrumentation.opentelemetry.io/inject-java: "true"` with both `instrumentation.opentelemetry.io/java-container-names` and `instrumentation.opentelemetry.io/nodejs-container-names`.

If I am reading this logic correctly the first use case is being blocked, is that correct?

Reply:
> Users should be allowed to enable the feature gate but still use instrumentation.opentelemetry.io/inject-java: "true" and provide no container names.

User is able to inject single instrumentation without *-container-names when `multi-instrumentation` feature gate is `enabled`.

> Users should not be allowed to enable the feature gate and use instrumentation.opentelemetry.io/inject-java: "true" with both instrumentation.opentelemetry.io/java-container-names and instrumentation.opentelemetry.io/nodejs-container-names.

I believe you were thinking about `instrumentation.opentelemetry.io/java-container-names` and `instrumentation.opentelemetry.io/container-names` - so in case of disabled `multi-instrumentation` only `container-names` annotation is used. When `multi-instrumentation` is enabled then specific for language container-names is used. There're tests in pod mutator related to this scenario.